src/librustc_driver/driver.rs

```diff
@@ -177,7 +163,32 @@ pub fn compile_input(sess: &Session,
             )?
         };
 
-        write_out_deps(sess, &outputs, &crate_name);
+        let output_paths = generated_output_paths(sess, &outputs, &crate_name);
+
+        // Ensure the source file isn't accidentally overwritten during compilation.
+        match *input_path {
```

Comment:
I'd personally use `if let Some(ref input_path) = *input_path { ... }` here, especially since the `None` branch is empty.

Reply:
(oh but I guess the pre-existing code that you're refactoring used `match`. Fine to leave it as is to ease reading the diff here, I guess.)

Reply:
It turns out the diffing algorithm doesn't catch on anyway, so I'll change it (considering the change is currently breaking on `x86_64-pc-windows-msvc`).